### ISSUE TITLE
torch.onnx.export Pad ops with ONNX opset 11

### ISSUE THREAD

**Issue by rmccorm4** [pytorch/pytorch#35516]
## 🐛 Bug

I don't know if this is a bug exactly, but between ONNX opset 10 and 11, there was a change to Pad ops making the `pads` an `input` to the node instead of an `attribute`.

This causes the ONNX graphs exported from PyTorch to be pretty complicated, and `torch.onnx.export(model, ..., do_constant_folding=True)` doesn't provide much help to simplify it.

However,[onnx-simplifier](https://github.com/daquexian/onnx-simplifier) does do a good job of simplifying the model back into a single Pad node.

The complex models generated by `torch.onnx.export(..., opset_version=11)` are causing some issues with the TensorRT ONNX parser.

There's one sample issue that discusses the problem a bit more, and that's where the sample code is below is taken from: https://github.com/NVIDIA/TensorRT/issues/439

## To Reproduce

1. Run the following code snippet with `OPSET=10`
2. Run the following code snippet with `OPSET=11`
3. Observe the difference between the generated ONNX graphs.

```python
import onnx
import argparse
import torch
import torch.nn as nn

OPSET = 10
# OPSET = 11

class MinimalModel(nn.Module):
    def __init__(self):
        super(MinimalModel, self).__init__()
        self.constant_zero_pad = nn.ConstantPad2d((1, 0, 0, 0), 0)

    def forward(self, input_tensor):
        return self.constant_zero_pad(input_tensor)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='PSMNet')
    parser.add_argument('output_onnx')
    args = parser.parse_args()

    minimal_model = MinimalModel()
    minimal_model = nn.DataParallel(minimal_model)
    minimal_model.cuda()

    # Random deep feature
    input_tensor = torch.rand((1, 32, 128, 128))
    # Check model can do a forward pass
    minimal_model(input_tensor)
    # Export to onnx
    torch.onnx.export(
        minimal_model.module,
        (input_tensor),
        args.output_onnx,
        export_params=True, verbose=True, training=False, opset_version=OPSET
    )

    original_model = onnx.load(args.output_onnx)
    onnx.checker.check_model(original_model)
```

## Expected behavior

I expect a simple ONNX graph with a INPUT -> PAD -> OUTPUT when exporting the above model with `torch.onnx.export(..., onnx_opset=11)`

## Environment

 - PyTorch Version (e.g., 1.0): 1.4
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6
 - CUDA/cuDNN version: 10.1/7.6
 - GPU models and configuration: V100
 - Any other relevant information: 

## Additional context

Model exported with ONNX Opset 10

![model-opset10](https://user-images.githubusercontent.com/21284872/77710946-c9cd4900-6f8c-11ea-93e0-0a8954f32697.PNG)

Model exported with ONNX Opset 11 + `do_constant_folding=True`

![model-constant-folding-opset11](https://user-images.githubusercontent.com/21284872/77710947-cb970c80-6f8c-11ea-8f4e-25af2612862f.PNG)

cc @suo @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by spandantiwari** [comment]
Just to clarify and confirm - is there something wrong you see with the export, or is it that the model export is correct but this is a question of multiple nodes being inserted in opset 11?

**Comment by rmccorm4** [comment]
Hi @spandantiwari ,

I believe the question is two-fold:

1. Why is the ONNX graph produced in Opset 11 so complicated vs. Opset 10
2. Why do additional flags like constant folding in `torch.onnx.export(...)` not seem to simplify the graph, but something like onnx-simplifier can achieve this: https://github.com/NVIDIA/TensorRT/issues/439#issuecomment-604163341?

**Comment by neginraoof** [comment]
Thanks for reporting this.
The difference between opset 10 and 11 export is because the spec of this operator has been updated in ONNX opset 11, and some op attributes have become op inputs. This updated has enabled export of pad operator with dynamic input shape in opset 11. You can export the model with pad op with an input tensor of certain shape and run it with another input of different shape.

But I agree that the opset 11 model can be optimized further (with constant folding). We'll look into this.

**Comment by neginraoof** [comment]
@rmccorm4 But again, as Spandan mentioned, please let us know if this issue is about model correctness or optimization.

**Comment by rmccorm4** [comment]
Hi @neginraoof ,

I believe the model is functionally correct, so I'd say optimization. But I'm not sure if the flags being used in torch.onnx.export are "correct" in handling this model since they don't seem to be doing as much as they could/should.

Thanks,
Ryan

**Comment by opeide** [comment]
In the example the padding is clearly constant though, is there any way to force the exporter to realize this? Is there any update on this or plan to fix? Exported models not being usable in tensorrt is pretty devastating, and the issue has been here for quite some time.

**Comment by wkl2013DeepVision** [comment]
same tool

ONNX IR version:  0.0.6
Opset version:    11
Producer name:    pytorch
Producer version: 1.7
Domain:
Model version:    0
Doc string:
----------------------------------------------------------------
[05/14/2021-09:03:53] [W] [TRT] onnx2trt_utils.cpp:220: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[05/14/2021-09:03:53] [W] [TRT] onnx2trt_utils.cpp:246: One or more weights outside the range of INT32 was clamped
ERROR: builtin_op_importers.cpp:2249 In function importPad:
[8] Assertion failed: inputs.at(1).is_weights()
Trying to use DLA core 0 on a platform that doesn't have any DLA cores
[05/14/2021-09:03:53] [E] [TRT] Network must have at least one output
[05/14/2021-09:03:53] [E] [TRT] Network validation failed.

**Comment by PeterL1n** [comment]
Also encountered this. The graph is so ugly even for a simple `nn.ZeroPad2d(2)`, which clearly has a constant padding. Please fix this.

**Comment by icyhearts** [comment]
similar problem. complicated model

**Comment by romil611** [comment]
i'm facing the issue. On running the following command:
` trtexec --onnx=efficientdet-d4.onnx --batch=1 --saveEngine=efficientdet-d4.trt --verbose`
i get 
`[06/10/2021-13:06:03] [V] [TRT] ModelImporter.cpp:125: Pad_53 [Pad] inputs: [x.1 -> (1, 3, 1024, 1024)], [1529 -> (-1)], [1530 -> ()], 
ERROR: builtin_op_importers.cpp:2179 In function importPad:
[8] Assertion failed: inputs.at(1).is_weights()
[06/10/2021-13:06:03] [E] Failed to parse onnx file
[06/10/2021-13:06:03] [E] Parsing model failed
[06/10/2021-13:06:03] [E] Engine creation failed
[06/10/2021-13:06:03] [E] Engine set up failed
&&&& FAILED TensorRT.trtexec # trtexec --onnx=efficientdet-d4.onnx --batch=1 --saveEngine=efficientdet-d4.trt --verbose
`

additional info:
`Input filename:   efficientdet-d4.onnx
ONNX IR version:  0.0.6
Opset version:    11
Producer name:    pytorch
Producer version: 1.7
Domain:           
Model version:    0
Doc string:       
`

`[06/10/2021-13:06:00] [I] === Model Options ===
[06/10/2021-13:06:00] [I] Format: ONNX
[06/10/2021-13:06:00] [I] Model: efficientdet-d4.onnx
[06/10/2021-13:06:00] [I] Output:
[06/10/2021-13:06:00] [I] === Build Options ===
[06/10/2021-13:06:00] [I] Max batch: 1
[06/10/2021-13:06:00] [I] Workspace: 16 MB
[06/10/2021-13:06:00] [I] minTiming: 1
[06/10/2021-13:06:00] [I] avgTiming: 8
[06/10/2021-13:06:00] [I] Precision: FP32
[06/10/2021-13:06:00] [I] Calibration: 
[06/10/2021-13:06:00] [I] Safe mode: Disabled
[06/10/2021-13:06:00] [I] Save engine: efficientdet-d4.trt
[06/10/2021-13:06:00] [I] Load engine: 
[06/10/2021-13:06:00] [I] Builder Cache: Enabled
[06/10/2021-13:06:00] [I] NVTX verbosity: 0
[06/10/2021-13:06:00] [I] Inputs format: fp32:CHW
[06/10/2021-13:06:00] [I] Outputs format: fp32:CHW
[06/10/2021-13:06:00] [I] Input build shapes: model
[06/10/2021-13:06:00] [I] Input calibration shapes: model
[06/10/2021-13:06:00] [I] === System Options ===
[06/10/2021-13:06:00] [I] Device: 0
[06/10/2021-13:06:00] [I] DLACore: 
[06/10/2021-13:06:00] [I] Plugins:
[06/10/2021-13:06:00] [I] === Inference Options ===
[06/10/2021-13:06:00] [I] Batch: 1
[06/10/2021-13:06:00] [I] Input inference shapes: model
[06/10/2021-13:06:00] [I] Iterations: 10
[06/10/2021-13:06:00] [I] Duration: 3s (+ 200ms warm up)
[06/10/2021-13:06:00] [I] Sleep time: 0ms
[06/10/2021-13:06:00] [I] Streams: 1
[06/10/2021-13:06:00] [I] ExposeDMA: Disabled
[06/10/2021-13:06:00] [I] Spin-wait: Disabled
[06/10/2021-13:06:00] [I] Multithreading: Disabled
[06/10/2021-13:06:00] [I] CUDA Graph: Disabled
[06/10/2021-13:06:00] [I] Skip inference: Disabled
[06/10/2021-13:06:00] [I] Inputs:
[06/10/2021-13:06:00] [I] === Reporting Options ===
[06/10/2021-13:06:00] [I] Verbose: Enabled
[06/10/2021-13:06:00] [I] Averages: 10 inferences
[06/10/2021-13:06:00] [I] Percentile: 99
[06/10/2021-13:06:00] [I] Dump output: Disabled
[06/10/2021-13:06:00] [I] Profile: Disabled
[06/10/2021-13:06:00] [I] Export timing to JSON file: 
[06/10/2021-13:06:00] [I] Export output to JSON file: 
[06/10/2021-13:06:00] [I] Export profile to JSON file: 
[06/10/2021-13:06:00] [I] 
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::GridAnchor_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::NMS_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::Reorg_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::Region_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::Clip_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::LReLU_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::PriorBox_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::Normalize_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::RPROI_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::BatchedNMS_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::FlattenConcat_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::CropAndResize version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::DetectionLayer_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::Proposal version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::ProposalLayer_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::PyramidROIAlign_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::ResizeNearest_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::Split version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::SpecialSlice_TRT version 1
[06/10/2021-13:06:00] [V] [TRT] Registered plugin creator - ::InstanceNormalization_TRT version 1
`

**Comment by ahangchen** [comment]
Any updates? I also encounter this problem, I can't use op10 to avoid it, because my model has other op that requires op11.
